src/portfft/descriptor_validate.hpp

```diff
@@ -0,0 +1,271 @@
+  // It helps to think of the 1D transform layed out in 2D with row length of stride, that way each element of a
+  // transform will be contiguous down a column.
+
+  // * If there is an index collision between batch N and batch N+M, then there will also be a collision between batch
+  // N-1 and batch N+M-1, so if there is any index collision, there will also be one with batch 0 (batch N-N and batch
+  // N+M-N).
+  // * If an index in a transform mod the stride of the transform is zero, then it would collide with the first batch,
+  // given an infinite FFT length. For all elements in a transforms, the index mod stride is the same.
+  // * If an element in a batch index collides with another batch, then all previous elements in that batch will also
+  // index collide with that batch, so we only need to check the first index of each batch.
+
+  const std::size_t fft_size = lengths[0];
+  const std::size_t stride = strides[0];
+
+  const std::size_t first_batch_limit = stride * fft_size;
+  if (first_batch_limit <= distance) {
+    return;
+  }
+
+  for (std::size_t b = 1; b < number_of_transforms;) {
+    std::size_t batch_first_idx = b * distance;
+    auto column = batch_first_idx % stride;
+    if (column == 0) {  // there may be a collision with the first batch
+      if (batch_first_idx >= first_batch_limit) {
+        // any further batch will only be further way
+        return;
+      }
+      throw invalid_configuration("Domain ", domain_str, ": batch ", b, " collides with first batch at index ",
+                                  batch_first_idx);
+    }
+
+    // there won't be another collision until the column number is near to stride again, so skip a few
+    auto batches_until_new_column = (stride - column) / distance;
+    if ((stride - column) % distance != 0) {
+      batches_until_new_column += 1;
+    }
+    b += batches_until_new_column;
+  }
+}
+
+/**
+ * Throw an exception if the given strides and distance are invalid for a single domain.
+ *
+ * @param lengths the dimensions of the tranform
+ * @param number_of_tranforms the number of batches
+ * @param strides the strides between elements in a domain
+ * @param distance the distance between batches in a domain
+ * @param domain_str a string with the name of the domain being validated
+ */
+inline void strides_distance_check(const std::vector<std::size_t>& lengths, std::size_t number_of_transforms,
+                                   const std::vector<std::size_t>& strides, std::size_t distance,
+                                   const std::string_view domain_str) {
+  validate_strides_distance_basic(lengths, number_of_transforms, strides, distance, domain_str);
+  if (lengths.size() > 1) {
+    strides_distance_multidim_check(lengths, number_of_transforms, strides, distance, domain_str);
+  } else {
+    strides_distance_1d_check(lengths, number_of_transforms, strides, distance, domain_str);
+  }
```

Comment:
Thank you :D
I believe the functionality of `strides_distance_multidim_check` is cover by these lines https://github.com/codeplaysoftware/portFFT/blob/471f48c3c30b1d5c643c41005dda70c01c5266d4/src/portfft/descriptor_validate.hpp#L164-L167.
I am confident `strides_distance_1d_check` is fast and covers all the 1D cases, but generalising that to multi-dimension was not quickly obvious to me, so just kept `strides_distance_multidim_check` to cover multi-dim cases.

Reply:
I see what you mean. The check in `strides_distance_multidim_check` is a little smarter than this `if` check. It sorts the indices by order of `generic_strides` so it can also quickly check that batch_interleaved configurations are valid. I think the equivalent would be:
```
const std::size_t first_batch_limit = stride * fft_size;
const std::size_t first_length_limit = distance * number_of_transforms;
if ((stride <= distance && first_batch_limit <= distance) ||
    (distance <= stride && first_length_limit <= stride)) {
    return;
}
```

Reply:
Are you sure that that is always correct?
I've added a bit to the loop so large batch interleaved layouts don't take long to check
https://github.com/codeplaysoftware/portFFT/blob/dee36f59096caa0830a422091fb5c09d8f8e7639/src/portfft/descriptor_validate.hpp#L181-L186

Reply:
I change my mind, you're correct, addressed here: https://github.com/codeplaysoftware/portFFT/pull/136/commits/c30f03068b9bdbfc8217642d2cbb1530bdaac25f